lib/net/protocol.rb

```diff
@@ -195,7 +195,10 @@ def readuntil(terminator, ignore_eof = false)
       offset = @rbuf_offset
       begin
         until idx = @rbuf.index(terminator, offset)
-          offset = @rbuf.bytesize
+          new_offset = @rbuf.bytesize - terminator.bytesize + 1
+          # Only assign the offset if it will advance.
+          # Otherwise an empty @rbuf could result in a negative offset.
```

Comment:
The fix here was to just set the offset in a way that it should only be set far enough ahead that a multiple byte terminator would still be read completely on the next read. For the comment, I'm not honestly sure if an empty buffer could be an issue but I played it safe and checked that new_offset is greater than offset before setting it.